core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java

```diff
@@ -0,0 +1,185 @@
+/****************************************************************************
+ * Copyright 2019, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+package com.optimizely.ab.optimizelyconfig;
+
+import com.optimizely.ab.annotations.VisibleForTesting;
+import com.optimizely.ab.config.*;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class OptimizelyConfigService {
+
+    private ProjectConfig projectConfig;
+    private OptimizelyConfig optimizelyConfig;
+
+    // Map containing variables list for every feature key used for merging variation and feature variables.
+    private Map<String, List<FeatureVariable>> featureKeyToVariablesMap;
```

Comment:
We do not need to keep this instance. Can we remove it after initialization? Same for projectConfig, which is needed for construction only.

Reply:
remove as in? assign null? I beleive whole `OptimizelyConfigService` instance will be garbage collected as we are not holding any reference to the service.

Reply:
To my understanding, your plan is to keep this OptimizelyConfigService instance for a cached copy of OptimizelyConfig.

Reply:
No, we are only keeping OptimizelyConfig instance for cache. i.e
`private final AtomicReference<OptimizelyConfig> currentOptimizelyConfig = new AtomicReference<>();`

Moreover, I agree with removing featureKeyToVariablesMap as an instance variable and generate it directly in the method where it is being used like that.

`Map<String, List<FeatureVariable>> featureKeyToVariablesMap = generateFeatureKeyToVariablesMap();`

`List<FeatureVariable> featureVariables = featureKeyToVariablesMap.get(featureKey);`

On the other hand I think we can keep projectConfig as an instance variable because it is being used in multiple methods and it's a clean approach to not send it to each method as an arguement.